File: Java/src/org/aachen/rpc/RequestHandler.java
package org.aachen.rpc;

import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Map;
import java.util.TreeMap;

import org.apache.xmlrpc.XmlRpcRequest;
import org.apache.xmlrpc.client.AsyncCallback;

public class RequestHandler {
	
	private static TreeMap<Integer, String> expectedRequestIps = new TreeMap<Integer, String>();
	private static TreeMap<Integer, String> differedRequestIps = new TreeMap<Integer, String>();
	private static LogicalClock localClock = new LogicalClock();
	private static ResourceHandler resourceHandler = new ResourceHandler();
	private static String classNameLog = "RequestHandler : ";
	
	private static String masterIp = JavaWsServer.getIpMaster();
	private static String myIp = JavaWsServer.getMyIpAddress();
	private static int myKey = JavaWsServer.getMyPriority();
	private static TreeMap<Integer, String> machines = JavaWsServer.getMachines();
	
	private static boolean haveInterest = false;
	private static boolean currentlyAccessing = false;
	private static boolean wantWrite = false;
	
	private static String finalString = "";
	private static String myString = "";
	
	/**
	 * class to handle async call back for ricart agrawala mutual exclusion
	 * @author ukimiawz
	 *
	 */
	private class CallBack implements AsyncCallback {
		private String classNameLog = "callBack Ricart Agrawala : ";

		@Override
		public void handleError(XmlRpcRequest arg0, Throwable arg1) {
			System.out.println(classNameLog + "Ricart Agrawala Async call failed");
		}

		@Override
		public void handleResult(XmlRpcRequest arg0, Object arg1) {
			System.out.println(classNameLog + "Ricart Agrawala Async call success");
		}
	}
	
	/***
	 * 
	 * @param wantWrite - indicates whether the request is for write or read
	 * @param isSignal - indicates whether the request coming from client or another server auto trigger
	 * @return
	 */
	public String startMessage(boolean wantWrite, boolean isSignal){
		
		try {
			System.out.println(classNameLog + "Wait for random amount of time");
			Thread.sleep((long)(Math.random() * 2000));
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		//initialize variables
		System.out.println(classNameLog + "Start mutual exclusion process");
		System.out.println(classNameLog + "Master IP =>" + masterIp);
		System.out.println(classNameLog + "My IP => " + myIp + " My key => " + myKey);
		
		if(!isSignal){
			//start signal from client, inform others
			//contact all machines to start
			TreeMap<Integer, String> machines = JavaWsServer.getMachines();
			System.out.println(classNameLog + "Contacting all nodes " + machines);
			Object[] params = new Object[]{wantWrite, true};
			XmlRpcHelper.SendToAllMachinesAsync(machines, "Request.startMessage", params, new CallBack());
		}
		
		System.out.println(classNameLog + "Initiating... Want to write => " + wantWrite);
		this.wantWrite = wantWrite;
		return sendRequest();
	}
	
	/***
	 * Start sending request to access shared resource
	 * @param wantWrite indicates whether the operation is write or read
	 * @return
	 */
	private String sendRequest(){
		
		//increment clock by 1
		localClock.incrementClock();
		System.out.println(classNameLog + "Lamport clock value incremented to " + localClock.getClockValue());
		
		//append all machines ip to list since we need to get permissions from all machine
		expectedRequestIps.clear();
		expectedRequestIps.putAll(machines);
		haveInterest = true;
		System.out.println(classNameLog + "Expecting permissions from " + expectedRequestIps);
		System.out.println(classNameLog + "My IP " + myIp);
		
		//request for permission to all machines
		for(Map.Entry<Integer,String> entry : machines.entrySet()) {
			
			String ipAddress = entry.getValue();
			int machineKey = entry.getKey();
			System.out.println(classNameLog + "Requesting permission from IP => " + ipAddress + " and machine key => " + machineKey);
			
			//remove me from request waiting list
			if(ipAddress == myIp){
				removeMachineFromExpected(machineKey);
				System.out.println(classNameLog + "Removing myself from machines list. My key " + machineKey);
			}
			
			try {
				//check if machine is on
				System.out.println(classNameLog + "Machine is alive. Asking permission from machine " + entry.getKey() + " => " + ipAddress);
				
				//ask permission to all
				String requestString = "Request.askPermission";
				Object[] params = new Object[]{localClock.getClockValue(), myKey, myIp, requestString};
				String replyOk = (String) XmlRpcHelper.SendToOneMachine(ipAddress, "Request.requestPermission", params);
				System.out.println(classNameLog + "Reply permission => " + replyOk);
				
				if(replyOk.equals("true") && replyOk.equals("false")){
					if(replyOk.equals("true")){
						System.out.println(classNameLog + "Got permission from " + machineKey + " removing machine from expected list");
						removeMachineFromExpected(machineKey);
					}
				} else {
					System.out.println(classNameLog + "Reply not as expected, removing machine from waiting list");
					removeMachineFromExpected(machineKey);
				}
				
			} catch (Exception e){
				System.out.println(classNameLog + "Machine is not valid, removing from expected list");
				removeMachineFromExpected(machineKey);
			}
		}
		
		System.out.println(classNameLog + "Finished requesting permission");
		
		if(expectedRequestIps.isEmpty()){
			System.out.println(classNameLog + "Everybody gave permission. Start resource access");
			String resource = doResourceAccess();
			System.out.println(classNameLog + "Resource value now " + resource);
		} else {
			System.out.println(classNameLog + "Wait until process finished");
			//wait until all finished
			while(haveInterest){
				try {
					System.out.println(".");
					Thread.sleep(500);
				} catch (InterruptedException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}
		
		if(wantWrite){
			//return written string
			System.out.println(classNameLog + "Written String => " + myString);
			return myString;
		}
		
		System.out.println(classNameLog + "Read String check my string => " + myString);
		System.out.println(classNameLog + "Final String => " + finalString);
		int containMyString = 0;
		
		if(finalString.contains(myString))
		{
			System.out.println(classNameLog + "!!!!My string is in the final string!!!!");
			containMyString = 1;
		}
		
		System.out.println(classNameLog + "Read process finished, final result => " + finalString + ";" + containMyString);
		return finalString + ";" + containMyString;
	}

	/***
	 * Receive message for ok permission, triggering resource access when all have given permission
	 * @param requestClock logical clock of incoming request
	 * @param machineKey key of machine sending ok signal
	 * @param ipAddress 
	 * @return string from resource access
	 */
	public String receivePermission(int requestClock, int machineKey, String ipAddress){
		
		//sync clock from message
		System.out.println(classNameLog + "Received permission");
		System.out.println(classNameLog + "Local clock " + localClock.getClockValue() + " request clock " + requestClock);
		localClock.syncClock(requestClock);
		System.out.println(classNameLog + "Local clock now " + localClock.getClockValue());
		System.out.println("Permission given from " + ipAddress);
		removeMachineFromExpected(machineKey);
		//check if all machine clear
		if(expectedRequestIps.isEmpty() && haveInterest){
			System.out.println("Permission clear, access resource");
			String nowResource;
			nowResource = doResourceAccess();
			System.out.println("Resource value now : " + nowResource);
		}
		
		return finalString;
	}

	/***
	 * Receive message asking permission from other machine
	 * @param requestClock logical clock of incoming request
	 * @param machineKey machine key of requesting machine
	 * @param ipAddress ip address of requesting machine
	 * @param requestString request string
	 * @return
	 */
	public String requestPermission(int requestClock, int machineKey, String ipAddress, String requestString){
		System.out.println(classNameLog + ipAddress + " asking for permission");
		//check if accessing resource
		if(!currentlyAccessing && !haveInterest){
			System.out.println(classNameLog + "I have no interest, go ahead");
			return "true";
		} else if(currentlyAccessing) {
			//add to differed ip
			System.out.println(classNameLog + "I'm currently accessing the resource, put request to waiting list");
			addMachineToDiffered(machineKey, ipAddress);
			System.out.println(classNameLog + differedRequestIps);
		} else if(haveInterest){
			System.out.println(classNameLog + "I'm currently in waiting list");
			if(requestClock < localClock.getClockValue()){
				System.out.println(classNameLog + "Request clock is earlier. You can go ahead before me");
				return "true";
			} else {
				System.out.println(classNameLog + "Request clock is later. Please line up after me");
				addMachineToDiffered(machineKey, ipAddress);
				System.out.println(classNameLog + differedRequestIps);
			}
		}
		
		//sync clock with clock from message
		System.out.println(classNameLog + "Local clock " + localClock.getClockValue() + " request clock " + requestClock);
		localClock.syncClock(requestClock);
		System.out.println(classNameLog + "Local clock now " + localClock.getClockValue());
		
		return "false";
	}
	
	/***
	 * Do resource access to master machine
	 * @return return string from master
	 */
	private String doResourceAccess(){
		System.out.println(classNameLog + "Accessing resource");
		currentlyAccessing = true;
		//clear permission, do request access
		if(wantWrite){
			System.out.println(classNameLog + "Writing to shared resource");
			myString = resourceHandler.generateRandomString();
			System.out.println(classNameLog + "Generated random string " + myString);
			finalString = resourceHandler.appendNewString(myIp, masterIp, myString);
			System.out.println(classNameLog + "Shared resource value now " + finalString);
		} else {
			finalString = resourceHandler.readNewString(myIp, masterIp); 
			System.out.println(classNameLog + "Reading resource with value " + finalString);
		}
		
		//send ok to all machines in differedRequestIps
		sendOkResponse();
		haveInterest = false;
		currentlyAccessing = false;
		System.out.println(classNameLog + "Flags cleared to false");
		
		return finalString;
	}
	
	/***
	 * Send ok signal to all machines in differed list so that they can remove this machine from expected list
	 */
	private void sendOkResponse(){
		//increment clock by 1
		System.out.println(classNameLog + "Sending signal that already finished accessing resource");
		System.out.println(classNameLog + differedRequestIps);
		localClock.incrementClock();
		System.out.println(classNameLog + "Local clock incremented to " + localClock.getClockValue());
		
		for(Map.Entry<Integer,String> entry : differedRequestIps.entrySet()) {
			String ipAddress = entry.getValue();
			int machineKey = entry.getKey();
			System.out.println(classNameLog + "Giving permission to " + ipAddress + " with machine number " + machineKey);
			
			Object[] params = new Object[]{localClock.getClockValue(), myKey, myIp};
			XmlRpcHelper.SendToOneMachineAsync(ipAddress, "Request.receivePermission", params, new CallBack());
			
			System.out.println(classNameLog + "Removing machine from differed request :" + machineKey);
			removeMachineFromDiffered(machineKey);
			System.out.println(classNameLog + differedRequestIps);
		}
	}
	
	/*==== Machine array management ====*/
	
	/***
	 * Remove machine from expected list
	 * @param machineKey machine key to be removed
	 */
	private void removeMachineFromExpected(int machineKey){
		System.out.println(classNameLog + "Remove machine key from expected request list => " + machineKey);
		if(expectedRequestIps.containsKey(machineKey)){
			expectedRequestIps.remove(machineKey);
		}
	}
	
	/***
	 * Add new machine to differed list
	 * @param machineKey machine key to be added
	 * @param ipAddress ip address of new machine to be added
	 */
	private void addMachineToDiffered(int machineKey, String ipAddress){
		System.out.println(classNameLog + "Add machine key to differed request list => " + machineKey);
		if(!differedRequestIps.containsKey(machineKey)){
			differedRequestIps.put(machineKey, ipAddress);
		}
	}
	
	/***
	 * Remove machine from differed list
	 * @param machineKey machine key to be removed
	 */
	private void removeMachineFromDiffered(int machineKey){
		System.out.println(classNameLog + "Remove machine key from differed request list => " + machineKey);
		if(differedRequestIps.containsKey(machineKey)){
			differedRequestIps.remove(machineKey);
		}
	}
}

File: C#/CSharp-XmlRpcNet/CSharpRpcServer/RequestHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


/// <summary>
/// Mutual Exclusion with Ricart
/// </summary>
public class RequestHandler
{
    private static Dictionary<int, String> expectedRequestIps = new Dictionary<int, String>();
    private static Dictionary<int, String> differedRequestIps = new Dictionary<int, string>();
    private static LogicalClock localClock = new LogicalClock();
    private static int timeout = 3000;
    private static ResourceHandler resourceHandler = new ResourceHandler();
    private static String classNameLog = "RequestHandler : ";

    private static String masterIp = CSharpRpcServer.getIpMaster();
    private static String myIp = CSharpRpcServer.getMyIpAddress();
    private static int myKey = CSharpRpcServer.getMyPriority();
    private static Dictionary<int, String> machines = CSharpRpcServer.getMachines();

    private static Boolean _haveInterest = false;
    private static Boolean _currentlyAccessing = false;
    private static Boolean _wantWrite = false;

    private static String _finalString = "";
    private static String _myString = "";

    public String startMessage(Boolean wantWrite, Boolean isSignal)
    {

        //initialize variables
        Console.WriteLine(classNameLog + "Start mutual exclusion process");
        Console.WriteLine(classNameLog + "Master IP =>" + masterIp);
        Console.WriteLine(classNameLog + "My IP => " + myIp + " My key => " + myKey);

        if (!isSignal)
        {
            //start signal from client, inform others
            //contact all machines to start
            Dictionary<int, String> machines = CSharpRpcServer.getMachines();
            Console.WriteLine(classNameLog + "Contacting all nodes " + machines);
            Object[] parameters = new Object[] { true };
            XmlRpcHelper.SendToAllMachinesAsync(machines, GlobalMethodName.requestHandlerStartMessage, parameters);
        }

        Console.WriteLine(classNameLog + "Initiating... Want to write => " + wantWrite);
        _wantWrite = wantWrite;
        return sendRequest();
    }

    //send request
    private String sendRequest()
    {

        //increment clock by 1
        localClock.incrementClock();
        Console.WriteLine(classNameLog + "Lamport clock value incremented to " + localClock.getClockValue());

        //append all machines ip to list since we need to get permissions from all machine
        expectedRequestIps.Clear();

        foreach (KeyValuePair<int, String> entry in machines)
        {
            expectedRequestIps.Add(entry.Key, entry.Value);
        }

        _haveInterest = true;
        Console.WriteLine(classNameLog + "Expecting permissions from " + expectedRequestIps);
        Console.WriteLine(classNameLog + "My IP " + myIp);

        //request for permission to all machines
        foreach (KeyValuePair<int, String> entry in machines)
        {

            String ipAddress = entry.Value;
            int machineKey = entry.Key;
            Console.WriteLine(classNameLog + "Requesting permission from IP => " + ipAddress + " and machine key => " + machineKey);

            //remove me from request waiting list
            if (ipAddress == myIp)
            {
                removeMachineFromExpected(machineKey);
                Console.WriteLine(classNameLog + "Removing myself from machines list. My key " + machineKey);
            }

            try
            {
                //check if machine is on
                if (NetworkHelper.isServerUp(ipAddress, 1090, 500))
                {
                    Console.WriteLine(classNameLog + "Machine is alive. Asking permission from machine " + entry.Key + " => " + ipAddress);

                    //ask permission to all
                    String requestString = "Request.askPermission";
                    Object[] parameters = new Object[] { localClock.getClockValue(), myKey, myIp, requestString };
                    String replyOk = (String)XmlRpcHelper.SendToOneMachine(ipAddress, "Request.requestPermission", parameters);
                    Console.WriteLine(classNameLog + "Reply permission => " + replyOk);

                    if (replyOk == "true" && replyOk == "false")
                    {
                        if (replyOk == "true")
                        {
                            Console.WriteLine(classNameLog + "Got permission from " + machineKey + " removing machine from expected list");
                            removeMachineFromExpected(machineKey);
                        }
                    }
                    else
                    {
                        Console.WriteLine(classNameLog + "Reply not as expected, removing machine from waiting list");
                        removeMachineFromExpected(machineKey);
                    }

                }
                else
                {
                    removeMachineFromExpected(machineKey);
                    Console.WriteLine(classNameLog + "Machine " + machineKey + " is not active. IP Address " + ipAddress);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Machine not valid {0}", e.Message);
                removeMachineFromExpected(machineKey);
                return e.Message;
            }
        }

        Console.WriteLine(classNameLog + "Finished requesting permission");

        if (expectedRequestIps.Count == 0)
        {
            Console.WriteLine(classNameLog + "Everybody gave permission. Start resource access");
            String resource = doResourceAccess();
            Console.WriteLine(classNameLog + "Resource value now " + resource);
        }
        else
        {
            Console.WriteLine(classNameLog + "Wait until process finished");
            //wait until all finished
            while (_haveInterest)
            {
                try
                {
                    Console.WriteLine(".");
                    System.Threading.Thread.Sleep(500);
                }
                catch (Exception e)
                {
                    Console.WriteLine("{0}", e.Message);
                    return e.Message;
                }
            }
        }

        if (_wantWrite)
        {
            //return written string
            Console.WriteLine(classNameLog + "Written String => " + _myString);
            return _myString;
        }

        Console.WriteLine(classNameLog + "Read String check my string => " + _myString);
        Console.WriteLine(classNameLog + "Final String => " + _finalString);
        int containMyString = 0;

        if (_finalString.Contains(_myString))
        {
            Console.WriteLine(classNameLog + "!!!!My string is in the final string!!!!");
            containMyString = 1;
        }

        Console.WriteLine(classNameLog + "Read process finished, final result => " + _finalString + ";" + containMyString);
        return _finalString + ";" + containMyString;
    }

    // receive message for ok permission, triggering resource access when all have given permission
    public String receivePermission(int requestClock, int machineKey, String ipAddress)
    {
        //sync clock from message
        Console.WriteLine(classNameLog + "Received permission");
        Console.WriteLine(classNameLog + "Local clock " + localClock.getClockValue() + " request clock " + requestClock);
        localClock.syncClock(requestClock);
        Console.WriteLine(classNameLog + "Local clock now " + localClock.getClockValue());
        Console.WriteLine("Permission given from " + ipAddress);
        removeMachineFromExpected(machineKey);
        //check if all machine clear
        if (expectedRequestIps.Count == 0 && _haveInterest)
        {
            Console.WriteLine("Permission clear, access resource");
            String nowResource;
            nowResource = doResourceAccess();
            Console.WriteLine("Resource value now : " + nowResource);
        }

        return _finalString;
    }

    //receive message request for permission
    public String requestPermission(int requestClock, int machineKey, String ipAddress, String requestString)
    {
        Console.WriteLine(classNameLog + ipAddress + " asking for permission");
        //check if accessing resource
        if (!_currentlyAccessing && !_haveInterest)
        {
            Console.WriteLine(classNameLog + "I have no interest, go ahead");
            return "true";
        }
        else if (_currentlyAccessing)
        {
            //add to differed ip
            Console.WriteLine(classNameLog + "I'm currently accessing the resource, put request to waiting list");
            addMachineToDiffered(machineKey, ipAddress);
            Console.WriteLine(classNameLog + differedRequestIps);
        }
        else if (_haveInterest)
        {
            Console.WriteLine(classNameLog + "I'm currently in waiting list");
            if (requestClock < localClock.getClockValue())
            {
                Console.WriteLine(classNameLog + "Request clock is earlier. You can go ahead before me");
                return "true";
            }
            else
            {
                Console.WriteLine(classNameLog + "Request clock is later. Please line up after me");
                addMachineToDiffered(machineKey, ipAddress);
                Console.WriteLine(classNameLog + differedRequestIps);
            }
        }

        //sync clock with clock from message
        Console.WriteLine(classNameLog + "Local clock " + localClock.getClockValue() + " request clock " + requestClock);
        localClock.syncClock(requestClock);
        Console.WriteLine(classNameLog + "Local clock now " + localClock.getClockValue());

        return "false";
    }

    private String doResourceAccess()
    {
        Console.WriteLine(classNameLog + "Accessing resource");
        _currentlyAccessing = true;
        //clear permission, do request access
        if (_wantWrite)
        {
            Console.WriteLine(classNameLog + "Writing to shared resource");
            _myString = resourceHandler.generateRandomString();
            Console.WriteLine(classNameLog + "Generated random string " + _myString);
            _finalString = resourceHandler.appendNewString(myIp, masterIp, _myString);
            Console.WriteLine(classNameLog + "Shared resource value now " + _finalString);
        }
        else
        {
            _finalString = resourceHandler.readNewString(myIp, masterIp);
            Console.WriteLine(classNameLog + "Reading resource with value " + _finalString);
        }

        //send ok to all machines in differedRequestIps
        sendOkResponse();
        _haveInterest = false;
        _currentlyAccessing = false;
        Console.WriteLine(classNameLog + "Flags cleared to false");

        return _finalString;
    }

    private void sendOkResponse()
    {
        //increment clock by 1
        Console.WriteLine(classNameLog + "Sending signal that already finished accessing resource");
        Console.WriteLine(classNameLog + differedRequestIps);
        localClock.incrementClock();
        Console.WriteLine(classNameLog + "Local clock incremented to " + localClock.getClockValue());

        foreach (KeyValuePair<int, String> entry in differedRequestIps)
        {
            String ipAddress = entry.Value;
            int machineKey = entry.Key;
            Console.WriteLine(classNameLog + "Giving permission to " + ipAddress + " with machine number " + machineKey);

            Object[] parameters = new Object[] { localClock.getClockValue(), myKey, myIp };
            Task<String> response = XmlRpcHelper.SendToOneMachineAsync(ipAddress, "Request.receivePermission", parameters);

            Console.WriteLine(classNameLog + "Removing machine from differed request :" + machineKey);
            removeMachineFromDiffered(machineKey);
            Console.WriteLine(classNameLog + differedRequestIps);
        }
    }

    #region Machine Array Management
    private void removeMachineFromExpected(int machineKey)
    {
        Console.WriteLine(classNameLog + "Remove machine key from expected request list => " + machineKey);
        if (expectedRequestIps.ContainsKey(machineKey))
        {
            expectedRequestIps.Remove(machineKey);
        }
    }

    private void addMachineToDiffered(int machineKey, String ipAddress)
    {
        Console.WriteLine(classNameLog + "Add machine key to differed request list => " + machineKey);
        if (!differedRequestIps.ContainsKey(machineKey))
        {
            differedRequestIps.Add(machineKey, ipAddress);
        }
    }

    private void removeMachineFromDiffered(int machineKey)
    {
        Console.WriteLine(classNameLog + "Remove machine key from differed request list => " + machineKey);
        if (differedRequestIps.ContainsKey(machineKey))
        {
            differedRequestIps.Remove(machineKey);
        }
    }
    #endregion
}

